src/main/java/com/cleanroommc/groovyscript/compat/mods/prodigytech/Solderer.java

```diff
@@ -0,0 +1,167 @@
+        return SoldererManager.RECIPES.removeIf(r -> {
+            if (r.requiresAdditive() && additive.test(r.getPattern())) {
+                addBackup(r);
+                return true;
+            }
+            return false;
+        });
+    }
+
+    @MethodDescription(example = @Example("item('prodigytech:circuit_refined')"))
+    public boolean removeByOutput(IIngredient output) {
+        return SoldererManager.RECIPES.removeIf(r -> {
+            if (output.test(r.getPattern())) {
+                addBackup(r);
+                return true;
+            }
+            return false;
+        });
+    }
+
+    @MethodDescription(priority = 2000, example = @Example(commented = true))
+    public void removeAll() {
+        SoldererManager.RECIPES.forEach(this::addBackup);
+        SoldererManager.removeAll();
+    }
+
+    @MethodDescription(priority = 2000, example = @Example(commented = true))
+    public void removeWithoutAdditive() {
+        SoldererManager.RECIPES.removeIf(r -> {
+            if (r.requiresAdditive()) return false;
+            addBackup(r);
+            return true;
+        });
+    }
+
+    @MethodDescription(type = MethodDescription.Type.QUERY)
+    public SimpleObjectStream<SoldererManager.SoldererRecipe> streamRecipes() {
+        return new SimpleObjectStream<>(SoldererManager.RECIPES)
+                .setRemover(this::remove);
+    }
+
+    @Property(property = "input", valid = {@Comp(type = Comp.Type.GTE, value = "0"), @Comp(type = Comp.Type.LTE, value = "1")})
+    @Property(property = "output", valid = @Comp("1"))
+    public static class RecipeBuilder extends AbstractRecipeBuilder<SoldererManager.SoldererRecipe> {
+
+        @Property(valid = @Comp(value = "1", type = Comp.Type.GTE))
+        private int time = Config.soldererProcessTime;
+
+        @Property(valid = @Comp(value = "1", type = Comp.Type.GTE))
+        private int gold;
+
+        @Property(valid = @Comp("1"))
+        private IIngredient pattern;
+
+        @RecipeBuilderMethodDescription
+        public Solderer.RecipeBuilder time(int time) {
+            this.time = time;
+            return this;
+        }
+
+        @RecipeBuilderMethodDescription
+        public Solderer.RecipeBuilder gold(int gold) {
+            this.gold = gold;
+            return this;
+        }
+
+        @RecipeBuilderMethodDescription
+        public Solderer.RecipeBuilder pattern(IIngredient pattern) {
+            this.pattern = pattern;
+            return this;
+        }
+
+        @Override
+        public String getErrorMsg() {
+            return "Error adding ProdigyTech Solderer Recipe";
+        }
+
+        @Override
+        public void validate(GroovyLog.Msg msg) {
+            validateItems(msg, 0, 1, 1, 1);
+            validateFluids(msg);
+            msg.add(gold <= 0, "gold must be greater than or equal to 1, yet it was {}", gold);
+            msg.add(pattern.isEmpty(), "pattern cannot be empty");

```

Comment:
use `IngredientHelper.isEmpty(pattern)` since it can be null

Reply:
pattern is nonnull in the mod code, although I guess I should still account for this in the builder to avoid leg shooting.